aws_signing_helper/pkcs11_signer.go

```diff
@@ -110,64 +270,157 @@ func getCerts(module *pkcs11.Ctx, session pkcs11.SessionHandle) (certs []*x509.C
 			return nil, err
 		}
 
 		rawCert := crtAttributes[0].Value
-		curCert, err := x509.ParseCertificate(rawCert)
+
+		var cert_obj CertObjInfo
+
+		cert_obj.x509, err = x509.ParseCertificate(rawCert)
 		if err != nil {
 			return nil, errors.New("error parsing certificate")
 		}
 
-		certs = append(certs, curCert)
+		// If we have any *additional* criteria in 'certIdentifier', drop
+		// any certificates which don't match. In the common case there
+		// not additional criteria and certMatches() matches everything.
+		// (Most implementations won't want that additional filter; in
+		// this code base it got grandfathered in).
+		//
+		// Fetch the CKA_ID and CKA_LABEL of the matching cert(s), so
+		// that they can be used later when hunting for the matching
+		// key.
+		if certMatches(certIdentifier, *cert_obj.x509) {
+			crtAttributes[0] = pkcs11.NewAttribute(pkcs11.CKA_ID, 0)
+			crtAttributes, err = module.GetAttributeValue(session, certObject, crtAttributes)
+			if err == nil {
+				cert_obj.id = crtAttributes[0].Value
+			}
+
+			crtAttributes[0] = pkcs11.NewAttribute(pkcs11.CKA_LABEL, 0)
+			crtAttributes, err = module.GetAttributeValue(session, certObject, crtAttributes)
+			if err == nil {
+				cert_obj.label = crtAttributes[0].Value
+			}
+
+			certs = append(certs, cert_obj)
+		}
 	}
 
 	return certs, nil
 }
 
-// Gets certificates that match the passed in CertIdentifier
-func GetMatchingPKCSCerts(certIdentifier CertIdentifier, lib string, slot uint) (module *pkcs11.Ctx, session pkcs11.SessionHandle, cert *x509.Certificate, matchingCerts []*x509.Certificate, err error) {
-	var certsFound []*x509.Certificate
+// Scan all matching slots to until we find certificates that match the uri
+// and additional criteria in certIdentifier.
+//
+// NB: This stops after it finds some. It's generally only looking for *one*
+// cert to use. If you want `p11tool --list-certificates`, use that instead.
+func getMatchingCerts(module *pkcs11.Ctx, slots []SlotIdInfo, certIdentifier CertIdentifier, uri *pkcs11uri.Pkcs11URI, single bool, pinPkcs11 *string) (session pkcs11.SessionHandle, slot_nr uint, logged_in bool, matchingCerts []CertObjInfo, err error) {
+	if uri != nil {
+		slots = matchSlots(slots, uri)
+	}
+
+	// http://david.woodhou.se/draft-woodhouse-cert-best-practice.html#rfc.section.8.1
+	//
+	// "For locating certificates, applications first iterate over the
+	// available tokens without logging in to them. In each token which
+	// matches the provided PKCS#11 URI, a search is performed for
+	// matching certificate objects. The first matching object is used
+	// as the certificate."
+	for _, slot := range slots {
+		session, err = module.OpenSession(slot.id, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKS_RO_PUBLIC_SESSION)
+		if err != nil {
+			continue
+		}
+
+		matchingCerts, err = getCertsInSession(module, session, uri, certIdentifier, single)
+		if err == nil && len(matchingCerts) > 0 {
+			return session, slot.id, false, matchingCerts, nil
```

Comment:
I think this is intentional, but we return before closing the session? My understanding is that the caller is responsible for closing the session after they're done using it. But in the case that there is a failure between the return of this function call and manually closing the session, would it be a small memory leak? I'm not sure what functionality Go provides for RAII exactly...

Reply:
We open a session, call `getCertsInSession()`, then if *successful* we return that session else we immediately close it. I don't think there's a leak there, is there? There can't be any other failure between the return of this call and closing the session.

Yes, there is a *potential* for leaks if the caller doesn't explicitly close the session, and it would be nice if we could make the `session` variable have a destructor which auto-closes it when it goes out of scope. I don't know the language well enough to make it do that; I settled for ensuring that we *do* always close it (I think) in the interesting case, although in the 'read-certificate-data` command I think we do deliberately leak it and don't care.